src/mds/query.py

```diff
@@ -64,22 +72,388 @@ async def search_metadata(
+        10, description="Maximum number of records returned. (max: 2000)"
+    ),
+    offset: int = Query(0, description="Return results at this given offset."),
+    #  XXX description
+    #  XXX how to name this python variable something other than filter but
+    #  still have client use "filter" as URL query param?
+    filter: str = Query("", description="Filters to apply."),
+):
+    """
+    Helper to search for metadata objects based on filters provided in filter
+    param.
+
+    The filtering functionality was primarily driven by the requirement that a
+    user be able to get all objects having an authz resource matching a
+    user-supplied pattern at any index in the "_resource_paths" array. For
+    example, given the following metadata objects:
+
+    {
+        "0": {
+            "message": "hello",
+            "_uploader_id": "100",
+            "_resource_paths": [
+                "/programs/a",
+                "/programs/b"
+            ],
+            "pet": "dog"
+        },
+        "1": {
+            "message": "greetings",
+            "_uploader_id": "101",
+            "_resource_paths": [
+                "/open",
+                "/programs/c/projects/a"
+            ],
+            "pet": "ferret",
+            "sport": "soccer"
+        },
+        "2": {
+            "message": "morning",
+            "_uploader_id": "102",
+            "_resource_paths": [
+                "/programs/d",
+                "/programs/e"
+            ],
+            "counts": [42, 42, 42],
+            "pet": "ferret",
+            "sport": "soccer"
+        },
+        "3": {
+            "message": "evening",
+            "_uploader_id": "103",
+            "_resource_paths": [
+                "/programs/f/projects/a",
+                "/admin"
+            ],
+            "counts": [1, 3, 5],
+            "pet": "ferret",
+            "sport": "basketball"
+        }
+    }
+
+    how do we design a filtering interface that allows the user to get all
+    objects having an authz string matching the pattern
+    "/programs/%/projects/%" at any index in its "_resource_paths" array? (%
+    has been used as the wildcard so far because that's what Postgres uses as
+    the wildcard for LIKE) In this case, the "1" and "3" objects should be
+    returned.
+
+    The filter syntax that was arrived at ending up following the syntax
+    specified by a Node JS implementation
+    (https://www.npmjs.com/package/json-api#filtering) of the JSON:API
+    specification (https://jsonapi.org/).
+
+    The format for this syntax is filter=(field_name,operator,value), in which
+    the field_name is a json key without quotes, operator is one of :eq, :ne,
+    :gt, :gte, :lt, :lte, :like, :all, :any (see operators dict), and value is
+    a typed json value against which the operator is run.
+
+    Examples:
+
+        - GET /objects?filter=(message,:eq,"morning") returns "2"
+        - GET /objects?filter=(counts.1,:eq,3) returns "3"
```

Comment:
the design doc doesn't use a `filter` param, and the syntax is more simple - we might want to accept `{record_or_metadata}.{arbitrary_key}={value}` as param like the design doc describes, and build the filter object based on that?

Reply:
This pull request still maintains the `{arbitrary_key}={value}` param syntax for the `GET /metadata` endpoint. I also don't think it would be too bad to support that syntax for the `GET /objects` endpoint in addition to the `filter=(counts.1,:eq,3)` style.